File: native/src/matrix.c
#include <include/matrix.h>

Matrix* matrix_new(int rows, int cols, MatrixType type) {
  Matrix* m = malloc(sizeof(Matrix));
  m->rows = rows;
  m->cols = cols;
  m->type = type;
  m->returned_to_js = 0;
  switch (type) {
    case TYPE_U32:
    case TYPE_I32:
    case TYPE_F32:
      // All of these types are 4 byte sized
      m->data = malloc(4 * rows * cols);
      memset(m->data, 0, 4 * rows * cols);
      break;
    default:
      free(m);
      return NULL;
  }
  return m;
}

Matrix* matrix_new_randf(int rows, int cols) {
  Matrix* m = malloc(sizeof(Matrix));
  m->rows = rows;
  m->cols = cols;
  m->type = TYPE_F32;
  m->returned_to_js = 0;
  float* data = malloc(4 * rows * cols);
  for (int i = 0; i < rows * cols; i++) {
    data[i] = gauss_rand();
  }
  m->data = (void*)data;
  return m;
}

Matrix* matrix_new_from_array(int rows, int cols, MatrixType type, void* data) {
  Matrix* m = matrix_new(rows, cols, type);
  if (m == NULL) {
    return NULL;
  }
  switch (type) {
    case TYPE_U32:
    case TYPE_I32:
    case TYPE_F32:
      memcpy(m->data, data, 4 * rows * cols);
      break;
    default:
      matrix_free(m);
      return NULL;
  }
  return m;
}

Matrix* matrix_new_from_array_zero_copy(int rows, int cols, MatrixType type, void* data) {
  Matrix* m = malloc(sizeof(Matrix));
  m->rows = rows;
  m->cols = cols;
  m->type = type;
  m->data = data;
  m->returned_to_js = 0;
  return m;
}

Matrix* matrix_new_fill_u32(int rows, int cols, uint32_t v) {
  Matrix* m = matrix_new(rows, cols, TYPE_U32);
  uint32_t* data = (uint32_t*) m->data;
  for (int i = 0; i < rows * cols; i++) {
    data[i] = v;
  }
  return m;
}

Matrix* matrix_new_fill_i32(int rows, int cols, int32_t v) {
  Matrix* m = matrix_new(rows, cols, TYPE_I32);
  int32_t* data = (int32_t*) m->data;
  for (int i = 0; i < rows * cols; i++) {
    data[i] = v;
  }
  return m;
}

Matrix* matrix_new_fill_f32(int rows, int cols, float v) {
  Matrix* m = matrix_new(rows, cols, TYPE_F32);
  float* data = (float*) m->data;
  for (int i = 0; i < rows * cols; i++) {
    data[i] = v;
  }
  return m;
}

Matrix* matrix_copy(Matrix* m) {
  Matrix* copy = matrix_new(m->rows, m->cols, m->type);
  switch (m->type) {
    case TYPE_U32:
    case TYPE_I32:
    case TYPE_F32:
      memcpy(copy->data, m->data, 4 * m->rows * m->cols);
      break;
    default:
      matrix_free(copy);
      return NULL;
  }
  return copy;
}

Matrix* matrix_map_f32(Matrix* m, float (*f) (float)) {
  if (m->type != TYPE_F32) return NULL;
  Matrix* result = matrix_new(m->rows, m->cols, m->type);
  float* resultData = result->data;
  float* mData = m->data;
  for (int i = 0; i < m->rows; i++) {
    for (int j = 0; j < m->cols; j++) {
      resultData[i * m->cols + j] = f(mData[i * m->cols + j]);
    }
  }
  return result;
}

Matrix* matrix_map_i32(Matrix* m, int32_t (*f) (int32_t)) {
  if (m->type != TYPE_I32) return NULL;
  Matrix* result = matrix_new(m->rows, m->cols, m->type);
  int32_t* resultData = result->data;
  int32_t* mData = m->data;
  for (int i = 0; i < m->rows; i++) {
    for (int j = 0; j < m->cols; j++) {
      resultData[i * m->cols + j] = f(mData[i * m->cols + j]);
    }
  }
  return result;
}

Matrix* matrix_map_u32(Matrix* m, uint32_t (*f) (uint32_t)) {
  if (m->type != TYPE_U32) return NULL;
  Matrix* result = matrix_new(m->rows, m->cols, m->type);
  uint32_t* resultData = result->data;
  uint32_t* mData = m->data;
  for (int i = 0; i < m->rows; i++) {
    for (int j = 0; j < m->cols; j++) {
      resultData[i * m->cols + j] = f(mData[i * m->cols + j]);
    }
  }
  return result;
}

Matrix* matrix_dot(Matrix* a, Matrix* b, Matrix* result) {
  if (a->type != b->type) return NULL;
  if (a->cols != b->rows) {
    printf("Matrix dimensions do not match for dot product\n");
    return NULL;
  }
  
  if (result == NULL) result = matrix_new(a->rows, b->cols, a->type);
  else {
    if (result->type != a->type) return NULL;
    if (result->rows != a->rows) return NULL;
    if (result->cols != b->cols) return NULL;
  }

  switch (a->type) {
    case TYPE_U32: {
      uint32_t* aData = a->data;
      uint32_t* bData = b->data;
      uint32_t* resultData = result->data;
      int i, j, k;
      #pragma omp parallel for private(i, j, k) shared(aData, bData, resultData) 
      for (i = 0; i < a->rows; i++) {
        for (j = 0; j < b->cols; j++) {
          uint32_t sum = 0;
          for (k = 0; k < a->cols; k++) {
            sum += aData[i * a->cols + k] * bData[k * b->cols + j];
          }
          resultData[i * b->cols + j] = sum;
        }
      }
      break;
    }

    case TYPE_I32: {
      int32_t* aData = a->data;
      int32_t* bData = b->data;
      int32_t* resultData = result->data;
      int i, j, k;
      #pragma omp parallel for private(i, j, k) shared(aData, bData, resultData) 
      for (i = 0; i < a->rows; i++) {
        for (j = 0; j < b->cols; j++) {
          int32_t sum = 0;
          for (k = 0; k < a->cols; k++) {
            sum += aData[i * a->cols + k] * bData[k * b->cols + j];
          }
          resultData[i * b->cols + j] = sum;
        }
      }
      break;
    }

    case TYPE_F32: {
      float* aData = a->data;
      float* bData = b->data;
      float* resultData = result->data;
      int i, j, k;
      #pragma omp parallel for private(i, j, k) shared(aData, bData, resultData) 
      for (i = 0; i < a->rows; i++) {
        for (j = 0; j < b->cols; j++) {
          float sum = 0;
          for (k = 0; k < a->cols; k++) {
            sum += aData[i * a->cols + k] * bData[k * b->cols + j];
          }
          resultData[i * b->cols + j] = sum;
        }
      }
      break;
    }

    default:
      matrix_free(result);
      return NULL;
  }
  return result;
}

Matrix* matrix_add(Matrix* a, Matrix* b, Matrix* result) {
  if (a->type != b->type) return NULL;
  if (a->rows != b->rows || a->cols != b->cols) return NULL;
  
  if (result == NULL) result = matrix_new(a->rows, a->cols, a->type);
  else {
    if (result->type != a->type) return NULL;
    if (result->rows != a->rows) return NULL;
    if (result->cols != a->cols) return NULL;
  }

  switch (a->type) {
    case TYPE_U32: {
      uint32_t* aData = a->data;
      uint32_t* bData = b->data;
      uint32_t* resultData = result->data;
      for (int i = 0; i < a->rows; i++) {
        for (int j = 0; j < a->cols; j++) {
          resultData[i * a->cols + j] = aData[i * a->cols + j] + bData[i * a->cols + j];
        }
      }
      break;
    }

    case TYPE_I32: {
      int32_t* aData = a->data;
      int32_t* bData = b->data;
      int32_t* resultData = result->data;
      for (int i = 0; i < a->rows; i++) {
        for (int j = 0; j < a->cols; j++) {
          resultData[i * a->cols + j] = aData[i * a->cols + j] + bData[i * a->cols + j];
        }
      }
      break;
    }

    case TYPE_F32: {
      float* aData = a->data;
      float* bData = b->data;
      float* resultData = result->data;
      for (int i = 0; i < a->rows; i++) {
        for (int j = 0; j < a->cols; j++) {
          resultData[i * a->cols + j] = aData[i * a->cols + j] + bData[i * a->cols + j];
        }
      }
      break;
    }

    default:
      matrix_free(result);
      return NULL;
  }
  return result;
}

Matrix* matrix_sub(Matrix* a, Matrix* b, Matrix* result) {
  if (a->type != b->type) return NULL;
  if (a->rows != b->rows || a->cols != b->cols) return NULL;
  
  if (result == NULL) result = matrix_new(a->rows, a->cols, a->type);
  else {
    if (result->type != a->type) return NULL;
    if (result->rows != a->rows) return NULL;
    if (result->cols != a->cols) return NULL;
  }

  switch (a->type) {
    case TYPE_U32: {
      uint32_t* aData = a->data;
      uint32_t* bData = b->data;
      uint32_t* resultData = result->data;
      for (int i = 0; i < a->rows; i++) {
        for (int j = 0; j < a->cols; j++) {
          resultData[i * a->cols + j] = aData[i * a->cols + j] - bData[i * a->cols + j];
        }
      }
      break;
    }

    case TYPE_I32: {
      int32_t* aData = a->data;
      int32_t* bData = b->data;
      int32_t* resultData = result->data;
      for (int i = 0; i < a->rows; i++) {
        for (int j = 0; j < a->cols; j++) {
          resultData[i * a->cols + j] = aData[i * a->cols + j] - bData[i * a->cols + j];
        }
      }
      break;
    }

    case TYPE_F32: {
      float* aData = a->data;
      float* bData = b->data;
      float* resultData = result->data;
      for (int i = 0; i < a->rows; i++) {
        for (int j = 0; j < a->cols; j++) {
          resultData[i * a->cols + j] = aData[i * a->cols + j] - bData[i * a->cols + j];
        }
      }
      break;
    }

    default:
      matrix_free(result);
      return NULL;
  }
  return result;
}

Matrix* matrix_add_f32(Matrix* a, float b) {
  if (a->type != TYPE_F32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  float* aData = a->data;
  float* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] + b;
    }
  }
  return result;
}

Matrix* matrix_sub_f32(Matrix* a, float b) {
  if (a->type != TYPE_F32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  float* aData = a->data;
  float* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] - b;
    }
  }
  return result;
}

Matrix* matrix_mul_f32(Matrix* a, float b) {
  if (a->type != TYPE_F32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  float* aData = a->data;
  float* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] * b;
    }
  }
  return result;
}

Matrix* matrix_div_f32(Matrix* a, float b) {
  if (a->type != TYPE_F32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  float* aData = a->data;
  float* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] / b;
    }
  }
  return result;
}

Matrix* matrix_add_u32(Matrix* a, uint32_t b) {
  if (a->type != TYPE_U32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  uint32_t* aData = a->data;
  uint32_t* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] + b;
    }
  }
  return result;
}

Matrix* matrix_sub_u32(Matrix* a, uint32_t b) {
  if (a->type != TYPE_U32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  uint32_t* aData = a->data;
  uint32_t* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] - b;
    }
  }
  return result;
}

Matrix* matrix_mul_u32(Matrix* a, uint32_t b) {
  if (a->type != TYPE_U32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  uint32_t* aData = a->data;
  uint32_t* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] * b;
    }
  }
  return result;
}

Matrix* matrix_div_u32(Matrix* a, uint32_t b) {
  if (a->type != TYPE_U32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  uint32_t* aData = a->data;
  uint32_t* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] / b;
    }
  }
  return result;
}

Matrix* matrix_add_i32(Matrix* a, int32_t b) {
  if (a->type != TYPE_I32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  int32_t* aData = a->data;
  int32_t* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] + b;
    }
  }
  return result;
}

Matrix* matrix_sub_i32(Matrix* a, int32_t b) {
  if (a->type != TYPE_I32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  int32_t* aData = a->data;
  int32_t* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] - b;
    }
  }
  return result;
}

Matrix* matrix_mul_i32(Matrix* a, int32_t b) {
  if (a->type != TYPE_I32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  int32_t* aData = a->data;
  int32_t* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] * b;
    }
  }
  return result;
}

Matrix* matrix_div_i32(Matrix* a, int32_t b) {
  if (a->type != TYPE_I32) return NULL;
  
  Matrix* result = matrix_new(a->rows, a->cols, a->type);
  int32_t* aData = a->data;
  int32_t* resultData = result->data;
  for (int i = 0; i < a->rows; i++) {
    for (int j = 0; j < a->cols; j++) {
      resultData[i * a->cols + j] = aData[i * a->cols + j] / b;
    }
  }
  return result;
}

Matrix* matrix_transpose(Matrix* m) {
  Matrix* result = matrix_new(m->cols, m->rows, m->type);
  switch (m->type) {
    case TYPE_U32: {
      uint32_t* mData = m->data;
      uint32_t* resultData = result->data;
      for (int i = 0; i < m->rows; i++) {
        for (int j = 0; j < m->cols; j++) {
          resultData[j * m->rows + i] = mData[i * m->cols + j];
        }
      }
      break;
    }

    case TYPE_I32: {
      int32_t* mData = m->data;
      int32_t* resultData = result->data;
      for (int i = 0; i < m->rows; i++) {
        for (int j = 0; j < m->cols; j++) {
          resultData[j * m->rows + i] = mData[i * m->cols + j];
        }
      }
      break;
    }

    case TYPE_F32: {
      float* mData = m->data;
      float* resultData = result->data;
      for (int i = 0; i < m->rows; i++) {
        for (int j = 0; j < m->cols; j++) {
          resultData[j * m->rows + i] = mData[i * m->cols + j];
        }
      }
      break;
    }

    default:
      matrix_free(result);
      return NULL;
  }
  return result;
}

void matrix_print(Matrix* m, char* name) {
  printf("%s = [\n", name);
  for (int i = 0; i < m->rows; i++) {
    printf("  ");
    for (int j = 0; j < m->cols; j++) {
      switch (m->type) {
        case TYPE_U32:
          printf("%u", ((uint32_t*)m->data)[i * m->cols + j]);
          break;
        case TYPE_I32:
          printf("%d", ((int32_t*)m->data)[i * m->cols + j]);
          break;
        case TYPE_F32:
          printf("%f", ((float*)m->data)[i * m->cols + j]);
          break;
        default:
          printf("?");
          break;
      }
      if (j < m->cols - 1) printf(", ");
    }
    printf("\n");
  }
  printf("]\n");
}

void matrix_serialize(Matrix* m, FILE* f) {
  fwrite(&m->type, sizeof(MatrixType), 1, f);
  fwrite(&m->rows, sizeof(uint32_t), 1, f);
  fwrite(&m->cols, sizeof(uint32_t), 1, f);
  fwrite(m->data, 4, m->rows * m->cols, f);
}

Matrix* matrix_deserialize(FILE* f) {
  MatrixType type;
  uint32_t rows, cols;
  fread(&type, sizeof(MatrixType), 1, f);
  fread(&rows, sizeof(uint32_t), 1, f);
  fread(&cols, sizeof(uint32_t), 1, f);
  Matrix* m = matrix_new(rows, cols, type);
  fread(m->data, 4, rows * cols, f);
  return m;
}

void matrix_free(Matrix* m) {
  CHECK_NULL(m);
  if (m->data != NULL) free(m->data);
  free(m);
}

File: backends/native/matrix.ts
import ffi from "./ffi.ts";

const {
  matrix_new,
  matrix_new_randf,
  matrix_new_from_array,
  matrix_new_fill_f32,
  matrix_new_fill_i32,
  matrix_new_fill_u32,
  matrix_free,
  matrix_dot,
  matrix_add,
  matrix_sub,
  matrix_add_f32,
  matrix_add_i32,
  matrix_add_u32,
  matrix_div_f32,
  matrix_div_i32,
  matrix_div_u32,
  matrix_copy,
  matrix_mul_f32,
  matrix_mul_i32,
  matrix_mul_u32,
  matrix_sub_f32,
  matrix_sub_i32,
  matrix_sub_u32,
  matrix_transpose,
} = ffi;

export type DataType = "u32" | "i32" | "f32";

enum C_DATA_TYPE {
  u32 = 0,
  i32 = 1,
  f32 = 2,
}

function typeFromArray(array: Float32Array | Int32Array | Uint32Array) {
  if (array instanceof Float32Array) return C_DATA_TYPE.f32;
  if (array instanceof Int32Array) return C_DATA_TYPE.i32;
  if (array instanceof Uint32Array) return C_DATA_TYPE.u32;
  throw new Error("Unsupported array type");
}

export type DataTypeArray<T extends DataType> = T extends "u32" ? Uint32Array
  : T extends "i32" ? Int32Array
  : T extends "f32" ? Float32Array
  : never;

const {
  op_ffi_read_u64,
  op_ffi_read_u32,
  op_ffi_read_u8,
  op_ffi_get_buf,
  // deno-lint-ignore no-explicit-any
} = (Deno as any).core.ops;

const MatrixFinalizer = new FinalizationRegistry((id: Deno.PointerValue) => {
  matrix_free(id);
});

export class Matrix<T extends DataType> {
  #ptr: Deno.PointerValue = 0;
  #token = { ptr: 0 as Deno.PointerValue };

  get unsafePointer() {
    return this.#ptr;
  }

  declare readonly type: T;
  declare readonly rows: number;
  declare readonly cols: number;
  declare readonly data: DataTypeArray<T>;

  constructor(ptr: Deno.PointerValue);
  constructor(rows: number, cols: number, type: T);
  constructor(rows: number, cols: number, data: DataTypeArray<T>);
  constructor(rows: number, cols: number, type: T, fill: number);
  constructor(
    rows: number,
    cols?: number,
    type?: T | DataTypeArray<T>,
    fill?: number | Deno.PointerValue,
  ) {
    if (cols === undefined) {
      this.#ptr = rows as Deno.PointerValue;
    } else if (typeof type === "string" && type in C_DATA_TYPE) {
      if (typeof fill === "number") {
        switch (type) {
          case "u32":
            this.#ptr = matrix_new_fill_u32(rows, cols!, fill);
            break;
          case "i32":
            this.#ptr = matrix_new_fill_i32(rows, cols!, fill);
            break;
          case "f32":
            this.#ptr = matrix_new_fill_f32(rows, cols!, fill);
            break;
        }
      } else {
        this.#ptr = matrix_new(rows, cols!, C_DATA_TYPE[type]);
      }
      Object.defineProperty(this, "type", { value: type, writable: false });
    } else if (typeof type === "object" && type !== null) {
      const ty = typeFromArray(type);
      Object.defineProperty(this, "type", {
        value: C_DATA_TYPE[ty],
        writable: false,
      });
      this.#ptr = matrix_new_from_array(rows, cols!, ty, type);
    } else throw new Error("Invalid arguments");

    const c = Number(this.#ptr);

    const ptr = op_ffi_read_u64(c);
    rows = op_ffi_read_u32(c + 8);
    cols = op_ffi_read_u32(c + 12);

    if (!("type" in this)) {
      const ty = op_ffi_read_u8(c + 16);
      Object.defineProperty(this, "type", {
        value: C_DATA_TYPE[ty],
        writable: false,
      });
    }

    const buf = op_ffi_get_buf(ptr, rows * cols! * 4);

    Object.defineProperties(this, {
      rows: { value: rows, writable: false },
      cols: { value: cols, writable: false },
      data: {
        value: this.type === "f32"
          ? new Float32Array(buf)
          : this.type === "i32"
          ? new Int32Array(buf)
          : new Uint32Array(buf),
        writable: false,
      },
    });

    this.#token.ptr = this.#ptr;
    MatrixFinalizer.register(this, this.#ptr, this.#token);
  }

  static rand(rows: number, cols: number): Matrix<"f32"> {
    return new Matrix(matrix_new_randf(rows, cols));
  }

  static of(data: number[][]): Matrix<"f32"> {
    const rows = data.length;
    const cols = data[0].length;
    return new Matrix(rows, cols, new Float32Array(data.flat()));
  }

  static row(data: number[]): Matrix<"f32"> {
    return new Matrix(1, data.length, new Float32Array(data));
  }

  static column(data: number[]): Matrix<"f32"> {
    return new Matrix(data.length, 1, new Float32Array(data));
  }

  dot(b: Matrix<T>): Matrix<T> {
    const c = matrix_dot(this.#ptr, b.unsafePointer);
    if (c === 0) throw new Error("Invalid matrix dimensions");
    return new Matrix(c);
  }

  add(b: Matrix<T> | number): Matrix<T> {
    if (typeof b === "number") {
      switch (this.type) {
        case "u32":
          return new Matrix(matrix_add_u32(this.#ptr, b));
        case "i32":
          return new Matrix(matrix_add_i32(this.#ptr, b));
        case "f32":
          return new Matrix(matrix_add_f32(this.#ptr, b));
        default:
          throw new Error("unreachable");
      }
    } else {
      const c = matrix_add(this.#ptr, b.unsafePointer);
      if (c === 0) throw new Error("Invalid matrix dimensions");
      return new Matrix(c);
    }
  }

  sub(b: Matrix<T> | number): Matrix<T> {
    if (typeof b === "number") {
      switch (this.type) {
        case "u32":
          return new Matrix(matrix_sub_u32(this.#ptr, b));
        case "i32":
          return new Matrix(matrix_sub_i32(this.#ptr, b));
        case "f32":
          return new Matrix(matrix_sub_f32(this.#ptr, b));
        default:
          throw new Error("unreachable");
      }
    } else {
      const c = matrix_sub(this.#ptr, b.unsafePointer);
      if (c === 0) throw new Error("Invalid matrix dimensions");
      return new Matrix(c);
    }
  }

  mul(b: Matrix<T> | number): Matrix<T> {
    if (typeof b === "number") {
      switch (this.type) {
        case "u32":
          return new Matrix(matrix_mul_u32(this.#ptr, b));
        case "i32":
          return new Matrix(matrix_mul_i32(this.#ptr, b));
        case "f32":
          return new Matrix(matrix_mul_f32(this.#ptr, b));
        default:
          throw new Error("unreachable");
      }
    } else {
      return this.dot(b);
    }
  }

  div(b: number): Matrix<T> {
    switch (this.type) {
      case "u32":
        return new Matrix(matrix_div_u32(this.#ptr, b));
      case "i32":
        return new Matrix(matrix_div_i32(this.#ptr, b));
      case "f32":
        return new Matrix(matrix_div_f32(this.#ptr, b));
      default:
        throw new Error("unreachable");
    }
  }

  transpose(): Matrix<T> {
    return new Matrix(matrix_transpose(this.#ptr));
  }

  copy(): Matrix<T> {
    return new Matrix(matrix_copy(this.#ptr));
  }

  free() {
    if (this.#ptr) {
      matrix_free(this.#ptr);
      MatrixFinalizer.unregister(this.#token);
      this.#ptr = 0;
    }
  }

  [Symbol.for("Deno.customInspect")]() {
    let result = `Matrix<${this.type}>(${this.rows}x${this.cols}) [\n`;
    for (let i = 0; i < this.rows; i++) {
      result += "  [" +
        this.data.subarray(i * this.cols, (i + 1) * this.cols).join(", ") +
        "]" + (i < this.rows - 1 ? "," : "") + "\n";
    }
    result += "]";
    return result;
  }
}
